cmd/zrepl/zrepl.c

```diff
@@ -734,6 +1075,137 @@ uzfs_zvol_mgmt_thread(void *arg)
+uzfs_zvol_rebuild_scanner_callback(off_t offset, size_t len,
+    blk_metadata_t *metadata, zvol_state_t *zv, void *args)
+{
+	zvol_io_hdr_t	hdr;
+	zvol_io_cmd_t	*zio_cmd;
+	zvol_rebuild_t  *warg;
+	zvol_info_t	*zinfo;
+
+	warg = (zvol_rebuild_t *)args;
+	zinfo = warg->zinfo;
+
+	hdr.version = REPLICA_VERSION;
+	hdr.opcode = ZVOL_OPCODE_READ;
+	hdr.io_seq = metadata->io_num;
+	hdr.offset = offset;
+	hdr.len = len;
+	hdr.flags = ZVOL_OP_FLAG_REBUILD;
+	hdr.status = ZVOL_OP_STATUS_OK;
+	printf("IO number for rebuild %ld\n", metadata->io_num);
+	zio_cmd = zio_cmd_alloc(&hdr, warg->fd);
+	/* Take refcount for uzfs_zvol_worker to work on it */
+	uzfs_zinfo_take_refcnt(zinfo, B_FALSE);
+	zio_cmd->zv = zinfo;
+	uzfs_zvol_worker(zio_cmd);
+	return (0);
+}
+
+/*
+ * Rebuild scanner function which after receiving
+ * vol_name and IO number, will scan metadata and
+ * read data and send across.
+ */
+static void
+uzfs_zvol_rebuild_scanner(void *arg)
+{
+	int		fd;
+	zvol_info_t	*zinfo = NULL;
+	zvol_io_hdr_t	hdr;
+	int 		rc = 0;
+	zvol_rebuild_t  warg;
+	char 		*name;
+	blk_metadata_t	metadata;
+	uint64_t	rebuild_req_offset;
+	uint64_t	rebuild_req_len;
+	zvol_io_cmd_t	*zio_cmd;
+
+	fd = *(int *)arg;
+	free(arg);
+
+read_socket:
+	rc = uzfs_zvol_read_header(fd, &hdr);
+	if (rc != 0) {
+		close(fd);
+		goto exit;
+	}
+
+	printf("op_code=%d io_seq=%ld\n", hdr.opcode, hdr.io_seq);
+	switch (hdr.opcode) {
+
+		case ZVOL_OPCODE_HANDSHAKE:
+			name = kmem_alloc(hdr.len, KM_SLEEP);
+			rc = uzfs_zvol_socket_read(fd, name, hdr.len);
+			if (rc != 0) {
+				free(name);
+				ZREPL_ERRLOG("Socket read error: %d\n", errno);
+				goto exit;
+			}
+
+			zinfo = uzfs_zinfo_lookup(name);
+			if (zinfo == NULL) {
+				ZREPL_ERRLOG("Volume/LUN: %s not found", name);
+				free(name);
+				goto exit;
+			}
+			free(name);
+			warg.zinfo = zinfo;
+			warg.fd = fd;
+			goto read_socket;
+			break;
+
+		case ZVOL_OPCODE_REBUILD_STEP:
+
+			metadata.io_num = hdr.checkpointed_io_seq;
+			rebuild_req_offset = hdr.offset;
+			rebuild_req_len = hdr.len;
+
+			ZREPL_LOG("Checkpointed IO_seq: %ld, "
+			    "Rebuild Req offset:%ld, Rebuild Req length:%ld\n",
+			    metadata.io_num, rebuild_req_offset,
+			    rebuild_req_len);
+
+			rc = uzfs_get_io_diff(zinfo->zv, &metadata,
+			    uzfs_zvol_rebuild_scanner_callback,
+			    rebuild_req_offset, rebuild_req_len, &warg);
+			if (rc != 0) {
+				printf("Rebuild scanning failed\n");
+			}
+			bzero(&hdr, sizeof (hdr));
+			hdr.status = ZVOL_OP_STATUS_OK;
+			hdr.version = REPLICA_VERSION;
+			hdr.opcode = ZVOL_OPCODE_REBUILD_STEP_DONE;
+			zio_cmd = zio_cmd_alloc(&hdr, fd);
+			/* Take refcount for uzfs_zvol_worker to work on it */
+			uzfs_zinfo_take_refcnt(zinfo, B_FALSE);
+			zio_cmd->zv = zinfo;
+			uzfs_zvol_worker(zio_cmd);
+			goto read_socket;
+			break;
+
+		case ZVOL_OPCODE_REBUILD_COMPLETE:
+			ZREPL_LOG("Rebuild process is over on Replica:%s\n",
+			    zinfo->name);
+			goto exit;
+			break;
+
+		default:
+			ASSERT(0);
```

Comment:
IMO the rule should be to use assert for internal inconsistencies (in scope of one zrepl process), but we should not crash zrepl because other zrepl or target sends us something unexpected. Better fit would be to log an error message and exit the thread here.